### PR TITLE
processor, sink: use global resolved-ts to flushSink() when got pipeline.CommandTypeStopAtTs

### PR DESCRIPTION
**Pull request by ben1009**
### What problem does this PR solve? 
fix CDC cloud: data inconsistency after CDC owner switching #2230, solution described in https://github.com/pingcap/ticdc/issues/2230#issuecomment-878272322 

### What is changed and how it works?
detail could be found in https://github.com/pingcap/ticdc/issues/2230#issuecomment-878272322 

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: cdc/processor/pipeline/sink.go [hunk 1]
```diff
@@ -269,12 +268,11 @@ func (n *sinkNode) Receive(ctx pipeline.NodeContext) error {
...
 	case pipeline.MessageTypeCommand:
 		if msg.Command.Tp == pipeline.CommandTypeStopAtTs {
-			if msg.Command.StoppedTs < n.checkpointTs {
-				log.Warn("the stopped ts is less than the checkpoint ts, "+
-					"the table pipeline can't be stopped accurately, will be stopped soon",
-					zap.Uint64("stoppedTs", msg.Command.StoppedTs), zap.Uint64("checkpointTs", n.checkpointTs))
-			}
 			n.targetTs = msg.Command.StoppedTs
+			n.barrierTs = msg.Command.StoppedTs
+			if err := n.flushSink(ctx, msg.Command.StoppedTs); err != nil {
```
Comment (amyangfei) [inline cdc/processor/pipeline/sink.go:273]: `n.flushSink(ctx, n.resolvedTs)` is called in other places, could this still flush more events beyond `global resolved ts`

File: cdc/processor/pipeline/sink.go [hunk 2]
```diff
@@ -269,12 +268,11 @@ func (n *sinkNode) Receive(ctx pipeline.NodeContext) error {
 		}
 	case pipeline.MessageTypeCommand:
 		if msg.Command.Tp == pipeline.CommandTypeStopAtTs {
-			if msg.Command.StoppedTs < n.checkpointTs {
-				log.Warn("the stopped ts is less than the checkpoint ts, "+
-					"the table pipeline can't be stopped accurately, will be stopped soon",
-					zap.Uint64("stoppedTs", msg.Command.StoppedTs), zap.Uint64("checkpointTs", n.checkpointTs))
-			}
 			n.targetTs = msg.Command.StoppedTs
+			n.barrierTs = msg.Command.StoppedTs
```
Comment (liuzix) [inline cdc/processor/pipeline/sink.go:272]: Why do we need to set `n.barrierTs`?

File: cdc/processor/pipeline/sink.go (hunk 1 above)
Comment (liuzix) [reply cdc/processor/pipeline/sink.go:273]: I believe that by design no event is ever flushed beyond `global resolved ts`.

File: cdc/processor/processor.go [hunk 3]
```diff
@@ -368,7 +368,7 @@ func (p *processor) handleTableOperation(ctx cdcContext.Context) error {
 				if opt.BoundaryTs < globalCheckpointTs {
 					log.Warn("the BoundaryTs of remove table operation is smaller than global checkpoint ts", zap.Uint64("globalCheckpointTs", globalCheckpointTs), zap.Any("operation", opt))
 				}
-				if !table.AsyncStop(opt.BoundaryTs) {
+				if !table.AsyncStop(p.changefeed.Status.ResolvedTs) {
```
Comment (liuzix) [inline cdc/processor/processor.go:371]: Why don't we modify the scheduler to ensure that the `global resolved ts` is used as the boundary to stop the table?

File: cdc/processor/pipeline/sink.go (hunk 2 above)
Comment (ben1009) [reply cdc/processor/pipeline/sink.go:272]: if not  
```
	if resolvedTs > n.barrierTs {
		resolvedTs = n.barrierTs
	}
```

https://github.com/pingcap/ticdc/blob/78987aa2b736994b86a31d1c0c5848226720f2b2/cdc/processor/pipeline/sink.go#L119

Comment (liuzix) [reply cdc/processor/pipeline/sink.go:272]: Is it possible that `resolvedTs > n.barrierTs`, if we don't manually set `n.barrierTs` explicit when stopping the table?

Comment (ben1009) [reply cdc/processor/pipeline/sink.go:272]: according to the code, yes, this is a chance
https://github.com/pingcap/ticdc/blob/2fc6415c04560ed1e6bfb159ff4960bf01b22c65/cdc/processor/processor.go#L646
```
// pushResolvedTs2Table sends global resolved ts to all the table pipelines.
func (p *processor) pushResolvedTs2Table() error {
	resolvedTs := p.changefeed.Status.ResolvedTs
	for _, table := range p.tables {
		table.UpdateBarrierTs(resolvedTs)
	}
	return nil
}
```
https://github.com/pingcap/ticdc/blob/2fc6415c04560ed1e6bfb159ff4960bf01b22c65/cdc/processor/pipeline/table.go#L95

File: cdc/processor/processor.go (hunk 3 above)
Comment (amyangfei) [inline cdc/processor/processor.go:371]: After `table.AsyncStop` is called, do we have a phase to wait the table checkpoint to reach the global resolved ts.

If not, in L388 the BoundaryTs of new table could be less than global resolved ts, and there still exists the risk that two processors write single table.

**Review by liuzix — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Comment by ben1009** [comment]
after the move table fail in ci, i run several tests simulate the move table operation in idc, all tests passed with the similar message `[2021/07/27 15:16:47.570 +08:00] [INFO] [main.go:69] ["check pass!!!"]`

File: cdc/processor/pipeline/sink.go [hunk 4]
```diff
@@ -132,6 +145,13 @@ func (n *sinkNode) flushSink(ctx pipeline.NodeContext, resolvedTs model.Ts) (err
 	if err != nil {
 		return errors.Trace(err)
 	}
+	log.Debug("flushSink.FlushRowChangedEvents",
+		zap.Uint64("resolvedTs", resolvedTs),
+		zap.Uint64("barrierTs", n.barrierTs),
+		zap.Uint64("targetTs", n.targetTs),
+		zap.Uint64("n.checkpointTs", n.checkpointTs),
```
Comment (overvenus) [inline cdc/processor/pipeline/sink.go:152]:
```suggestion
```

File: cdc/processor/pipeline/sink.go [hunk 5]
```diff
@@ -104,10 +104,16 @@ func (n *sinkNode) flushSink(ctx pipeline.NodeContext, resolvedTs model.Ts) (err
...
 			n.status.Store(TableStatusStopped)
+			log.Debug("set TableStatusStopped cause err != nil", zap.Error(err))
+
 			return
 		}
 		if n.checkpointTs >= n.targetTs {
 			n.status.Store(TableStatusStopped)
+			log.Debug("set TableStatusStopped cause n.checkpointTs >= n.targetTs",
+				zap.Uint64("n.checkpointTs", n.checkpointTs),
+				zap.Uint64("n.targetTs", n.targetTs))
```
Comment (overvenus) [inline cdc/processor/pipeline/sink.go:115]:
```suggestion
				zap.Uint64("checkpointTs", n.checkpointTs),
				zap.Uint64("targetTs", n.targetTs))
```

**Comment by ben1009** [comment]
Closed for now, since this fix introduce another issue if global resolved-ts fallback when moving table

**Pull request closed without merging**